RecoMET/METProducers/src/HcalNoiseInfoProducer.cc

```diff
@@ -500,9 +707,130 @@ HcalNoiseInfoProducer::filldigis(edm::Event& iEvent, const edm::EventSetup& iSet
 	      }
           } // end of HBHE check
      } // loop on HcalCalibDigiCollection
+
+     // now match the laser monitor data by fiber (in time) 
+     if( fillLaserMonitor_ ) {
+       // check for any fibers without data and fill
+       // them so we dont run into problems later
+       for( unsigned idx = 0; idx < laserMonDetTypeList_.size(); ++idx ) {
+           if( lasmon_adcs[idx].empty() ) {
+               lasmon_adcs[idx] = std::vector<int>(10, -1);
+           }
+           if( lasmon_capids[idx].empty() ) {
+               lasmon_capids[idx] = std::vector<int>(10, -1);
+           }
+       }
+       unsigned nFibers = laserMonIEtaList_.size();
+       // for each fiber we need to find the index at with the 
+       // data from the next fiber matches in order to stitch them together.
+       // When there is an overlap, the data from the end of the
+       // earlier fiber is removed.  There is no removal of the last fiber
+       std::vector<unsigned> matching_idx; 
+       // we assume that the list of fibers was given in time order
+       // (if this was not the case, then we just end up using 
+       // all data from all fibers )
+       for( unsigned fidx = 0; fidx < (nFibers - 1); ++fidx ) {
+         // start with the last TS and loop backwards
+         // on each iteration check if all capId and ADCs from this
+         // TS forward match the beginning entries of 
+         // the next fiber
+         int last_ts = lasmon_capids[fidx].size()-1;  // last TS
+         int start_ts = last_ts; // start_ts will be decrimented on each loop
+         // in the case that our stringent check below doesn't work 
+         // store the latest capID that has a match
+         int latest_cap_match = -1;
+         // reverse loop over TSs
+         while( start_ts >= 0 ) {
+           int ncheck = 0; // count the number of TS that were checked
+           int nmatch_cap = 0; // count the number of TS where capID matched
+           int nmatch_adc = 0; // count the number of TS where ADC matched
+
+           for( int cidx = start_ts, nidx = 0; cidx <= last_ts; cidx++, nidx++ ) { 
```

Comment:
At this point, also cidx and nidx could be unsigned

Reply:
nidx can be unsigned, but since start_ts needs to be allowed to go negative cidx should be an int.  I changed nidx to be unsigned

Reply:
start_ts is kept >= 0 by the while statement at L743

Reply:
the while statement only executes if start_ts >=0.  start_ts is decremented within the while loop if there is no match found.  in the case when the loop starts with start_ts == 0, it can therefore go to -1.  This is a special case which indicates that no match was found between two channels and is handled after the while loop

Reply:
When cidx is assigned start_ts, such a start_ts is not yet decremented.
Therefore cidx can't be assigned something below 0 by construction

Reply:
right, the indices themselves do not go below 0, but start_ts and last_ts should be signed to be consistent with their implementation.  therefore I can do this,

  for( unsigned cidx = (unsigned)start_ts, nidx=0; cidx <= (unsigned)last_ts; cidx++, nidx++ ) {

is this satisfactory?

Reply:
Not nice, indeed... On the other hand, also having a few indexes signed and a few unsigned in not so nice. About the performance I don't think it is an issue anyhow.

@davidlange6 : what do you suggest?

If there are no other suggestions, I will start the test with what will be in this PR in a couple of hours (as I will be mostly online in the meanwhile...)

Reply:
i had missed that - one could use a different "default" value - eg, std::numeric_limits<unsigned int>::max()

Reply:
sorry - ignore that - I think I didn't follow completely the logic yet

Reply:
I went ahead and updated it to what I suggested.  Let me know if you have other suggestions off of the most recent pull

Reply:
thinking outloud - doing

unsigned int tmp=0;
while ( tmp < last_ts+1 ) {
   unsigned int start_ts = last_ts-i;
 ..
}
if ( tmp > last_ts)...

would remove all the mixups between int and unsigned int. I didn't try to implement it so maybe its more effort than benefit. 

That said, this logic has a lot of places where "off-by-one" errors can come in. Its good to review it to see if there is a simpler approach.